automated/linux/perf/perf.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/sh
+
+# shellcheck disable=SC1091
+. ../../lib/sh-test-lib
+OUTPUT="$(pwd)/output"
+RESULT_FILE="${OUTPUT}/result.txt"
+export RESULT_FILE
+SKIP_INSTALL="true"
+# List of test cases
+TEST="record report stat test"
+# PERF version
+PERF_VERSION="$(uname -r | cut -d . -f 1-2)"
+
+usage() {
+    echo "Usage: $0 [-s <true|false>]" 1>&2
+    exit 1
+}
+
+while getopts "s:h" arg; do
+   case "$arg" in
+     s) SKIP_INSTALL="${OPTARG}";;
+     h|*) usage ;;
+   esac
+done
+
+# Run perf record tests
+run_perf_record() {
+    # Test 'perf record'
+    info_msg "Performing perf record test..."
+    TCID="perf_record_test"
+    perf record -e cycles -o perf-lava-test.data ls -a  2>&1 | tee perf-record.log
+    samples=$(grep -ao "[0-9]\\+[ ]\\+samples" perf-record.log| cut -f 1 -d' ')
+    if [ "${samples}" -gt 1 ]; then
+        report_pass "${TCID}"
+    else
+        report_fail "${TCID}"
+    fi
+    rm perf-record.log
+}
+
+# Run perf report tests
+run_perf_report() {
+    # Test 'perf report'
+    info_msg "Performing perf report test..."
+    TCID="perf_report_test"
+    perf report -i perf-lava-test.data 2>&1 | tee perf-report.log
+    pcnt_samples=$(grep -c -e "^[ ]\\+[0-9]\\+.[0-9]\\+%" perf-report.log)
+    if [ "${pcnt_samples}" -gt 1 ]; then
+        report_pass "${TCID}"
+    else
+        report_fail "${TCID}"
+    fi
+    rm perf-report.log perf-lava-test.data
+}
+
+# Run perf stat tests
+run_perf_stat() {
+    # Test 'perf stat'
+    info_msg "Performing perf stat test..."
+    TCID="perf_stat_test"
+    perf stat -e cycles ls -a 2>&1 | tee perf-stat.log
+    cycles=$(grep -o "[0-9,]\\+[ ]\\+cycles" perf-stat.log | sed 's/,//g' | cut -f 1 -d' ')
+    if [ -z "${cycles}" ]; then
+        report_skip "${TCID}"
+    else
+        if [ "${cycles}" -gt 1 ]; then
+            report_pass "${TCID}"
+        else
+            report_fail "${TCID}"
+        fi
+    fi
+    rm perf-stat.log
+}
+
+# Run perf test tests
+run_perf_test() {
+    # Test 'perf test'
+    info_msg "Performing 'perf test'..."
+    perf test -v
+    report_pass "perf_test"
+}
+
+# Test run.
+! check_root && error_msg "This script must be run as root"
+create_out_dir "${OUTPUT}"
+
+info_msg "About to run perf test..."
+info_msg "Output directory: ${OUTPUT}"
+
+if [ "${SKIP_INSTALL}" = "True" ] || [ "${SKIP_INSTALL}" = "true" ]; then
+    info_msg "install perf skipped"
+else
+    dist_name
+    # shellcheck disable=SC2154
+    case "${dist}" in
+      debian|ubuntu)
+	pkgs="linux-tools-${PERF_VERSION}"
+        install_deps "${pkgs}" "${SKIP_INSTALL}"
+        ;;
+      centos|fedora)
+	pkgs="perf"
```

Comment:
how do you control version of perf on centos?

Reply:
I do remember the fact that. perf version matching with kernel version is always a challenge.
Currently Open embedded recipe is getting pre-built the one i have tested is 4.20 is working.

Reply:
@mwasilew I have not tested perf install on centos.
I have checked on redhat page and found yum install perf is the way to install
# yum install perf 
ref link,
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/7/html/tuning_guide/about_perf

Reply:
@nareshkamboju my point here is that for debian based systems you try to install perf package matching running kernel. For RPM systems you install version that is 'latest'. IIRC yum can forcibly install version you require: yum install <package name>-<version info>

Reply:
I just checked in centos docker container. We can install particular package version in the format Milosz mentioned above.  You will need to check on centos for detecting kernel or package version. Within the container, perf package version doesn't match my host kernel, that is expected.

```
[root@ccf8f43bc871 /]# yum list perf --showduplicates
Loaded plugins: fastestmirror, ovl
Loading mirror speeds from cached hostfile
 * base: mirrors.aliyun.com
 * extras: mirrors.aliyun.com
 * updates: mirrors.aliyun.com
Available Packages
perf.x86_64                                                                                                  3.10.0-957.el7                                                                                                       base   
perf.x86_64                                                                                                  3.10.0-957.1.3.el7                                                                                                   updates
perf.x86_64                                                                                                  3.10.0-957.5.1.el7                                                                                                   updates
[root@ccf8f43bc871 /]# yum install perf-3.10.0-957.el7
```